### DISCUSSION TITLE
How to use RankTwoTensor within MOOSE? Tried following the documentation but I'm still having issues.

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#24210] [Q&A General] [answered]
I have a custom material class where I want to calculate some values from the strain tensor. I calculated the strain tensor with the "ADComputeSmallStrain" material class and fed it into mine.

The strain tensor can be accessed as ` _strain[_qp]` , correct? I have used components of the strain in other parts of the code, and they all seem to be Real/Doubles. For example, computing the 1st invariant:

`ADReal I_1 = _strain[_qp](0,0) + _strain[_qp](1,1) + _strain[_qp](2,2); `

works fine.

Now I want to calculate the principal strains, which can be gotten from the eigenvalues of the strain tensor. I found this documentation for working with RankTwoTensors:

https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html#ad2be11b8c319023932c59f2b09d330b1

Here's where the confusion begins. The header describes the function as:

void RankTwoTensorTempl< T >::symmetricEigenvalues (std::vector< T > & eigvals) const

That is, a function with only one argument. How is the symmetricEigenvalues function supposed to compute the eigenvalues without me even passing the tensor to it?

Next, an example is given as:

`{
  RankTwoTensorTempl<T> a;
  symmetricEigenvaluesEigenvectors (eigvals, a);
}`

where you actually do pass the tensor to it and it edits the vector in place, I assume, with the eigenvalues. This function has a slightly different name, though. I tried it, and couldn't get it to work because of yet another problem- the function does not recognize _strain[_qp] as a RankTwoTensor. 

I tried what I could find to convert it to a rank two tensor, even initializing the individual components like in the documentation:

`RankTwoTensor A(1, 2, 3, 4, 5, 6, 7, 8, 9);
// A = [ 1 4 7
//       2 5 8
//       3 6 9 ]`

i.e,

`RankTwoTensor A(_strain[_qp](0,0), ...)`

but that does not seem to work. I also noticed that when I tried to intialize a RankTwoTensor, other methods mentioned in the documentation were not found. For example, A.square() to find the square of the tensor.

Can someone please point me in the right direction on how to do tensor algebra within MOOSE? Thanks.

**Comment by hugary1995** [comment]
> I have a custom material class where I want to calculate some values from the strain tensor. I calculated the strain tensor with the "ADComputeSmallStrain" material class and fed it into mine.
> 
> The strain tensor can be accessed as ` _strain[_qp]` , correct? I have used components of the strain in other parts of the code, and they all seem to be Real/Doubles. For example, computing the 1st invariant:
> 
> `ADReal I_1 = _strain[_qp](0,0) + _strain[_qp](1,1) + _strain[_qp](2,2); `
> 
> works fine.

Correct. More concisely,
```
ADReal I_1 = _strain[_qp].trace();
```

> 
> Now I want to calculate the principal strains, which can be gotten from the eigenvalues of the strain tensor. I found this documentation for working with RankTwoTensors:
> 
> https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html#ad2be11b8c319023932c59f2b09d330b1
> 
> Here's where the confusion begins. The header describes the function as:
> 
> void RankTwoTensorTempl< T >::symmetricEigenvalues (std::vector< T > & eigvals) const
> 
> That is, a function with only one argument. How is the symmetricEigenvalues function supposed to compute the eigenvalues without me even passing the tensor to it?

This is C++ basic. See tutorials on C++ classes, for example this one https://cplusplus.com/doc/tutorial/classes/
What you need is
```
std::vector<ADReal> eigvals;
A.symmetricEigevalues(eigvals);
```

> 
> Next, an example is given as:
> 
> `{ RankTwoTensorTempl<T> a; symmetricEigenvaluesEigenvectors (eigvals, a); }`
> 
> where you actually do pass the tensor to it and it edits the vector in place, I assume, with the eigenvalues. This function has a slightly different name, though. I tried it, and couldn't get it to work because of yet another problem- the function does not recognize _strain[_qp] as a RankTwoTensor.

That is not an example. You misread the doxygen documentation. The quoted code below the method definition is its source code.

> 
> I tried what I could find to convert it to a rank two tensor, even initializing the individual components like in the documentation:
> 
> `RankTwoTensor A(1, 2, 3, 4, 5, 6, 7, 8, 9); // A = [ 1 4 7 // 2 5 8 // 3 6 9 ]`
> 
> i.e,
> 
> `RankTwoTensor A(_strain[_qp](0,0), ...)`
> 
> but that does not seem to work. I also noticed that when I tried to intialize a RankTwoTensor, other methods mentioned in the documentation were not found. For example, A.square() to find the square of the tensor.

That should work. Did you get a compiler error, or is it giving you wrong results? If it is a compiler error, the error messages are usually sufficient to help you figure out what's wrong. You can also post the compiler errors here, which is usually a better way than just saying "that does not seem to work".

> 
> Can someone please point me in the right direction on how to do tensor algebra within MOOSE? Thanks.

**Reply by richmondodufisan** [reply]
Thanks for the quick response. I realize I should've included an error message

` error: no matching function for call to 'RankTwoTensorTempl<double>::RankTwoTensorTempl(const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&, const MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true>&)'
  108 |      _strain[_qp](2,0), _strain[_qp](2,1), _strain[_qp](2,2));`

Apparently, there is an issue with the "type". I keep seeing something about expecting a double instead of a DualNumber. Do I need to individually convert each component to a double first?

Also, thank you for clarifying the documentation to me. I see now that the issue is most likely conversion between types.

**Reply by richmondodufisan** [reply]
My confusion also comes from trying simpler methods and letting it figure out the type itself with "auto". For example:

`auto strain_square = _strain[_qp].square();`

based on the documentation (https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html#afc410261c1e9a68ffa4ef8721bec7669) gave the error:

`error: 'MooseADWrapper<RankTwoTensorTempl<double>, true>' {aka 'const class RankTwoTensorTempl<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true> >'} has no member named 'square'
   83 |  auto strain_square = _strain[_qp].square();`

I  needed to compute the square of the tensor earlier in the code, but ended up doing it manually with the components. I'd love to be able to figure out what's wrong.

Also, I don't know if I missed it, but I couldn't find the documentation for a simple matrix multiplication, i.e A_ij * B_jk. 

I was thinking to simply do _strain[_qp] * _strain[_qp], but wasn't sure.

**Reply by hugary1995** [reply, marked as answer]
The error message suggests that you are trying to construct a non AD `RankTwoTensor` from `ADReal`s. If your material uses automatic differentiation, it is usually a good idea to make everything AD, e.g., `ADReal`, `ADRealVectorValue`, `ADRankTwoTensor`, `ADRankFourTensor`, etc., unless you are _absolutely_ sure that something will never need AD.

Your second error message (on `square`) is a weird one though. Either the compiler is lying, or you have an outdated MOOSE whose `RankTwoTensorTempl` does not offer a method called `square`. You can check the version of your MOOSE using `git rev-parse HEAD`. Alternatively, open the file `RankTwoTensor.h` and check if there is a definition of `square()` which looks like
```cpp
  RankTwoTensorTempl<T> square() const;
```

**Reply by hugary1995** [reply]
> Also, I don't know if I missed it, but I couldn't find the documentation for a simple matrix multiplication, i.e A_ij * B_jk.
> 
> I was thinking to simply do _strain[_qp] * _strain[_qp], but wasn't sure.

That's not your fault. The `RankTwoTensor` class inherits from `TypeTensor` in libMesh, see
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1TypeTensor.html

Some basic operators are inherited from there.

**Reply by hugary1995** [reply]
Also, if you click on the `operator*()` in the RankTwoTensor doxygen, it should automatically send you to the right method (in libMesh): https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1TypeTensor.html#a3f5f5939e09b4b07ecfc9d3ba2c4fc15

**Reply by richmondodufisan** [reply]
So I tried to update moose and unearthed another can of worms. Followed the instructions for updating (https://mooseframework.inl.gov/getting_started/new_users.html) and it is compiling with several errors, most of which look like this:

`/home/richmondodufisan/projects/moose/framework/build/header_symlinks/MooseUtils.h:457:46: error: 'TensorTraits' is not a member of 'libMesh::TensorTools'
  457 |     else if constexpr (libMesh::TensorTools::TensorTraits<T>::rank == 1)
      |                                              ^~~~~~~~~~~~
/home/richmondodufisan/projects/moose/framework/build/header_symlinks/MooseUtils.h:457:60: error: expected primary-expression before '>' token
  457 |     else if constexpr (libMesh::TensorTools::TensorTraits<T>::rank == 1)
      |                                                            ^
/home/richmondodufisan/projects/moose/framework/build/header_symlinks/MooseUtils.h:457:63: error: '::rank' has not been declared; did you mean 'std::rank'?
  457 |     else if constexpr (libMesh::TensorTools::TensorTraits<T>::rank == 1)`

Is it possible to revert to the previous version of MOOSE that worked fine? What could be the problem here?

---------------------------------------------------------------------------------------------

Regarding the RankTwoTensor issues, I see- the problem was switching between doubles and ADReals. If I get you right, the default for RankTwoTensor is a non-AD variable, so to initialize an AD Tensor, would I need to do something like this?

`RankTwoTensor<ADReal> A (1, 2, ...)`

I would've tested it but for the issues with MOOSE itself now.

**Reply by hugary1995** [reply]
If you are using conda, you're probably just missing a `conda update --all` or `mamba update --all`. After which you should `make clobberall` and re`make`. @GiudGiud may have other suggestions on the updating issue.

Use `RankTwoTensor` for non AD tensor, and `ADRankTwoTensor` for AD tensor. You do not need to provide the template parameter which is included in the typedefs, i.e.
```
typedef RankTwoTensorTempl<Real> RankTwoTensor;
typedef RankTwoTensorTempl<ADReal> ADRankTwoTensor;
```

**Reply by richmondodufisan** [reply]
Gotcha. Thanks! I did the conda update but it still isn't compiling (same errors). I also tried `git clean -xfd` and redid `make clobberall` and `make -j4` but it still has the same errors, so I'll wait on @GiudGiud . Thank you so much for your help!

**Reply by GiudGiud** [reply]
what does `mamba list | grep moose` return ?

**Reply by richmondodufisan** [reply]
I got 

`Command 'mamba' not found, did you mean:

  command 'samba' from deb samba (2:4.13.14+dfsg-0ubuntu0.20.04.4)

Try: sudo apt install <deb name>`

Do I need to install something? The last time I updated MOOSE on this particular computer was probably in late 2021

**Reply by GiudGiud** [reply]
umm.
well `conda install mamba` then `mamba update --all`

but keep in mind, if you update libmesh, then you will need to update MOOSE as well

**Reply by richmondodufisan** [reply]
In order, I did:

`conda install mamba`

`mamba update --all`

went to moose directory, then did

`git clean -xfd`

then reupdated moose with:

`git fetch origin` and

`git rebase origin/master`

but was told it was already up to date. Then I went to my app directory, and did

`make clobberall` and

`make -j4`

but I'm still getting the same errors. What else could be the problem?

**Reply by richmondodufisan** [reply]
$ mamba list | grep moose
 packages in environment at /home/richmondodufisan/miniconda3/envs/moose:
moose-libmesh             2021.07.14              build_0    idaholab
moose-libmesh-vtk         6.3.0                   build_7    idaholab
moose-mpich               3.3.2                   build_7    idaholab
moose-petsc               3.15.1                  build_2    idaholab
moose-tools               2021.01.06                 py_0    idaholab

Is there any other thing I need to do for these to update?

**Reply by GiudGiud** [reply]
moose-libmesh 2021.07.14 build_0 idaholab

really old still

At this point, your conda environment is frozen due to dependencies that are frozen in time.
The only solution is to remove the environment and install afresh.

If this fails, if `mamba list | grep moose` still returns 2021 instead of April 2023 for moose-libmesh, then remove conda, and install mamba.
Basically follow the first step there https://mooseframework.inl.gov/getting_started/installation/conda.html
after having deleted conda

**Reply by richmondodufisan** [reply]
I uninstalled and reinstalled everything and it works now. Thanks!
